### PR TITLE
NAS-117363 / 22.12 / NAS-117363: Splitting VDev into proper subtypes

### PR DESCRIPTION
**Pull request by undsoft**
Fixes a long running issue in our with VDev type being responsible for two things at the same type: a vdev and a disk within a vdev.
This renames `VDev` to `TopologyItem` and introduces additional subtypes.
Adds other minor changes.

Testing:
Regression testing for Devices..

Known issues:
Title icon in device details card is not fully implemented.

### CODE REVIEW

**Review by RehanY147 — APPROVED** [review]
No regressions

File: src/app/pages/storage/volumes/volume-status/volume-status.component.ts [hunk 1]
```diff
@@ -365,11 +365,11 @@ export class VolumeStatusComponent implements OnInit, AfterViewInit {
       _.find(actions, { id: 'offline' }).isHidden = true;
     }
 
-    if (vdevType === VDevType.Mirror || vdevType === VDevType.Replacing || vdevType === VDevType.Spare) {
+    if (vdevType === TopologyItemType.Mirror || (vdevType as any) === 'REPLACING' || vdevType === TopologyItemType.Spare) {
```
Comment (RehanY147) [inline src/app/pages/storage/volumes/volume-status/volume-status.component.ts:368]: We don't wanna use magic strings

File: src/app/pages/storage2/modules/devices/components/devices/devices.component.ts [hunk 2]
```diff
@@ -73,39 +77,42 @@ export class DevicesComponent implements OnInit {
...
       return flattenTreeWithFilter(dataNodes, (dataNode) => {
-        if (isVDev(dataNode)) {
-          switch (dataNode.type) {
-            case VDevType.Disk:
-              return dataNode.disk?.toLowerCase().includes(query.toLowerCase());
-            case VDevType.Mirror:
-              return dataNode.name?.toLowerCase().includes(query.toLowerCase());
-          }
-        } else {
+        if (isVdevGroup(dataNode)) {
```
Comment (RehanY147) [inline src/app/pages/storage2/modules/devices/components/devices/devices.component.ts:80]: Previously we were checking for `'stat' in obj` for this check which we're not doing in this code block now. Just want to make sure we're changing the outcome of this conditional block.

File: src/app/pages/storage2/modules/devices/components/disk-details-panel/disk-details-panel.component.ts [hunk 3]
```diff
@@ -8,7 +8,19 @@ import { Disk, VDev } from 'app/interfaces/storage.interface';
...
+
+  get title(): string {
+    if (isTopologyDisk(this.topologyItem)) {
+      return this.topologyItem.disk || this.topologyItem.guid;
+    }
+
+    return this.topologyItem.type;
+  }
+
+  get isVdevDisk(): boolean {
```
Comment (RehanY147) [inline src/app/pages/storage2/modules/devices/components/disk-details-panel/disk-details-panel.component.ts:23]: This looks redundant

**Review by RehanY147 — CHANGES_REQUESTED** [review]

File: src/app/pages/storage2/modules/devices/components/disk-details-panel/disk-details-panel.component.ts (hunk 3 above)
Comment (undsoft) [reply src/app/pages/storage2/modules/devices/components/disk-details-panel/disk-details-panel.component.ts:23]: Used in template, but I'll rename it.

File: src/app/pages/storage2/modules/devices/components/devices/devices.component.ts (hunk 2 above)
Comment (undsoft) [reply src/app/pages/storage2/modules/devices/components/devices/devices.component.ts:80]: The outcome is the same, changes names of functions and uses return early.

**Review by RehanY147 — APPROVED** [review]
Looks good

**Pull request merged**